src/mood.ts

```diff
@@ -0,0 +1,227 @@
+        (ensureTurns - haveEffect(effect)) / turnsPerCast(this.skill)
+      );
+    }
+    return haveEffect(effect) > ensureTurns;
+  }
+}
+
+class PotionMoodElement extends MoodElement {
+  potion: Item;
+  maxPricePerTurn: number;
+
+  constructor(potion: Item, maxPricePerTurn: number) {
+    super();
+    this.potion = potion;
+    this.maxPricePerTurn = maxPricePerTurn;
+  }
+
+  execute(ensureTurns: number): boolean {
+    // FIXME: Smarter buying logic.
+    // FIXME: Allow constructing stuff (e.g. snow cleats)
+    const effect = effectModifier(this.potion, "Effect");
+    const effectTurns = haveEffect(effect);
+    const turnsPerUse = numericModifier(this.potion, "Effect Duration");
+    if (mallPrice(this.potion) > this.maxPricePerTurn * turnsPerUse)
+      return false;
+    if (effectTurns < ensureTurns) {
+      // print(`${effect}: going for ${turns} turns, currently ${effectTurns}`);
+      const uses = (ensureTurns - effectTurns) / turnsPerUse;
+      const quantityToBuy = clamp(uses - availableAmount(this.potion), 0, 100);
+      buy(quantityToBuy, this.potion, this.maxPricePerTurn * turnsPerUse);
+      const quantityToUse = clamp(uses, 0, availableAmount(this.potion));
+      use(quantityToUse, this.potion);
+    }
+    return haveEffect(effect) >= ensureTurns;
+  }
+}
+
+class GenieMoodElement extends MoodElement {
+  effect: Effect;
+
+  constructor(effect: Effect) {
+    super();
+    this.effect = effect;
+  }
+
+  execute(ensureTurns: number): boolean {
+    if (haveEffect(this.effect) >= ensureTurns) return true;
+    const neededWishes = Math.ceil(
+      (haveEffect(this.effect) - ensureTurns) / 20
+    );
+    const wishesToBuy = clamp(
+      neededWishes - availableAmount($item`pocket wish`),
+      0,
+      20
+    );
+    buy(wishesToBuy, $item`pocket wish`, 50000);
+    let wishesToUse = clamp(
+      neededWishes,
+      0,
+      availableAmount($item`pocket wish`)
+    );
+    for (; wishesToUse > 0; wishesToUse--) {
+      cliExecute(`genie effect ${this.effect.name}`);
+    }
+    return haveEffect(this.effect) >= ensureTurns;
+  }
+}
+
+class CustomMoodElement extends MoodElement {
+  effect: Effect;
+  gainEffect: () => void;
+
+  constructor(effect: Effect, gainEffect?: () => void) {
+    super();
+    this.effect = effect;
+    this.gainEffect = gainEffect ?? (() => cliExecute(effect.default));
+  }
+
+  execute(ensureTurns: number): boolean {
+    let currentTurns = haveEffect(this.effect);
+    let lastCurrentTurns = -1;
+    while (currentTurns < ensureTurns && currentTurns !== lastCurrentTurns) {
+      this.gainEffect();
+      lastCurrentTurns = currentTurns;
+      currentTurns = haveEffect(this.effect);
+    }
+    return haveEffect(this.effect) > ensureTurns;
+  }
+}
+
+/**
+ * Class representing a mood object. Set options using the static methods, and add mood elements using the instance methods.
+ */
+export class Mood {
+  static songSlots: Effect[][] = [];
+  static useSausages = true;
```

Comment:
This should probably be an optional argument to the Mood constructor. And done via that in general rather than modifying statics surely?

Reply:
yeah, will tweak this in the next PR. I want to be able to set defaults for a whole script, e.g. for a script in-run to not use sausages in any of its moods.